include/deal.II/base/task_result.h

```diff
@@ -455,9 +455,18 @@ namespace Threads
   {
     // If we have waited before, then return immediately:
     if (result_is_available)
-      return;
+      {
+        std::lock_guard<std::mutex> lock(mutex);
+
+        // First make clear that the result is no longer available, then
+        // reset the object:
+        result_is_available = false;
+        task_result.reset();
+
+        return;
+      }
     else
-      // If we have not waited, wait now. We need to use the double-checking
+      // Otherwise, check again under a lock. We need to use the double-checking
       // pattern to ensure that if two threads get to this place at the same
       // time, one returns right away while the other does the work. Note
       // that this happens under the lock, so only one thread gets to be in
```

Comment:
If this is the intention (i.e., "ensure that if two threads get to this place at the same time, one returns right away while the other does the work") then the logic is still faulty:
```
if (result_is_available)
  {
    std::lock_guard<std::mutex> lock(mutex);
    // cleanup
  }
else
  {
    std::lock_guard<std::mutex> lock(mutex);
    if (result_is_available)
      {
        // cleanup
      }
    else
      {
        // something else
      }
  }
```
The cleanup should probably read:
```
if (result_is_available)
  {
    // Schmidt's double checking to ensure only one thread does cleanup:
    std::lock_guard<std::mutex> lock(mutex);
    if (result_is_available)
      {
        result_is_available = false;
        // cleanup
      }
  }
```

I do not understand why you want to check `result_is_available` again under a lock a second time. What does this accomplish?

Reply:
You're of course totally right!